### ISSUE TITLE
Errors packaging and validating wrapper profiles having dependencies

### ISSUE THREAD

**Issue by tpetchel** [inspec/inspec#2041]
### Description

While specifying dependencies from a wrapper profile, I can use `url`, but not the other fields, to package and validate my profile.

* 👍 `url`
* 🙅‍♂️ `git`
* 🙅‍♂️ `github`
* 🙅‍♂️ `compliance`

### InSpec and Platform Version

```
$ inspec version
1.31.1
```

macOS 10.12.4
Chef Automate 1.5.46  (using Chef Automate Pilot)

### Replication Case

* Run `inspec init profile` to create a profile.

```
$ inspec init profile issue-2041
Create new profile at /Users/thomaspetchel/scratch/profiles/issue-2041
 * Create directory controls
 * Create file controls/example.rb
 * Create file inspec.yml
 * Create directory libraries
 * Create file README.md
 * Create file libraries/.gitkeep
```

* Modify `inspec.yml` like each of the examples that follow.

### `url`

```
name: issue-2041
title: InSpec Profile
maintainer: The Authors
copyright: The Authors
copyright_email: you@example.com
license: Apache-2.0
summary: An InSpec Compliance Profile
version: 0.1.0
depends:
  - name: apache-baseline
    url: https://github.com/dev-sec/apache-baseline/archive/master.tar.gz
    #git: https://github.com/dev-sec/apache-baseline.git
    #github: dev-sec/apache-baseline
    #compliance: admin/apache-baseline
```

`inspec vendor` passes.

```
$ inspec vendor issue-2041 --overwrite
Dependencies for profile issue-2041 successfully vendored to issue-2041/vendor
```

`inspec check` also succeeds.

```
$ inspec check issue-2041
`command(httpd).exist?` is not suported on your OS:
Location:    issue-2041
Profile:     issue-2041
Controls:    2
Timestamp:   2017-08-01T13:24:10-04:00
Valid:       true

No errors or warnings
```

### `git`

```
name: issue-2041
title: InSpec Profile
maintainer: The Authors
copyright: The Authors
copyright_email: you@example.com
license: Apache-2.0
summary: An InSpec Compliance Profile
version: 0.1.0
depends:
  - name: apache-baseline
    #url: https://github.com/dev-sec/apache-baseline/archive/master.tar.gz
    git: https://github.com/dev-sec/apache-baseline.git
    #github: dev-sec/apache-baseline
    #compliance: admin/apache-baseline
```

`inspec vendor` succeeds.

```
$ inspec vendor issue-2041 --overwrite
Dependencies for profile issue-2041 successfully vendored to issue-2041/vendor
```

But now `inspec check` fails.

```
$ inspec check issue-2041
/Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/profile.rb:63:in `for_path': Don't understand inspec profile in /Users/thomaspetchel/.inspec/cache/dc2239efdbe6ed3d638d1f06ceb561891ed03ba0, it doesn't look like a supported profile structure. (RuntimeError)
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/profile.rb:72:in `for_fetcher'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/requirement.rb:106:in `profile'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/dependency_set.rb:65:in `block in each'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/dependency_set.rb:64:in `each'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/dependency_set.rb:64:in `each'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/profile.rb:169:in `load_libraries'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/profile.rb:453:in `load_checks_params'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/profile.rb:447:in `load_params'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/profile.rb:141:in `params'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/profile.rb:307:in `controls_count'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/profile.rb:278:in `check'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/cli.rb:69:in `check'
	from /opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/thor-0.19.1/lib/thor/command.rb:27:in `run'
	from /opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/thor-0.19.1/lib/thor/invocation.rb:126:in `invoke_command'
	from /opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/thor-0.19.1/lib/thor.rb:359:in `dispatch'
	from /opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/thor-0.19.1/lib/thor/base.rb:440:in `start'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/bin/inspec:12:in `<top (required)>'
	from /opt/chefdk/bin/inspec:95:in `load'
	from /opt/chefdk/bin/inspec:95:in `<main>'
```

### `github`

```
name: issue-2041
title: InSpec Profile
maintainer: The Authors
copyright: The Authors
copyright_email: you@example.com
license: Apache-2.0
summary: An InSpec Compliance Profile
version: 0.1.0
depends:
  - name: apache-baseline
    #url: https://github.com/dev-sec/apache-baseline/archive/master.tar.gz
    #git: https://github.com/dev-sec/apache-baseline.git
    github: dev-sec/apache-baseline
    #compliance: admin/apache-baseline
```

`inspec vendor` fails.

```
$ inspec vendor issue-2041 --overwrite
/Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/cached_fetcher.rb:15:in `initialize': Could not fetch inspec profile in {:cwd=>"issue-2041", :backend=>Inspec::Backend::Class @transport=Train::Transports::Mock::Connection, :name=>"apache-baseline", :github=>"dev-sec/apache-baseline"}. (RuntimeError)
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/requirement.rb:84:in `new'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/requirement.rb:84:in `fetcher'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/requirement.rb:62:in `resolved_source'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/resolver.rb:67:in `block in resolve'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/resolver.rb:59:in `each'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/resolver.rb:59:in `resolve'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/resolver.rb:31:in `resolve'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/dependency_set.rb:89:in `vendor'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/profile.rb:393:in `generate_lockfile'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/profile_vendor.rb:58:in `vendor_dependencies'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/profile_vendor.rb:15:in `vendor!'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/base_cli.rb:176:in `vendor_deps'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/cli.rb:114:in `vendor'
	from /opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/thor-0.19.1/lib/thor/command.rb:27:in `run'
	from /opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/thor-0.19.1/lib/thor/invocation.rb:126:in `invoke_command'
	from /opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/thor-0.19.1/lib/thor.rb:359:in `dispatch'
	from /opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/thor-0.19.1/lib/thor/base.rb:440:in `start'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/bin/inspec:12:in `<top (required)>'
	from /opt/chefdk/bin/inspec:95:in `load'
	from /opt/chefdk/bin/inspec:95:in `<main>'
```

(Not checking `inspec check` because `vendor` fails...)

### `compliance`

(You need a Chef Automate installation and to have run `inspec compliance login_automate` to test this one.)

```
name: issue-2041
title: InSpec Profile
maintainer: The Authors
copyright: The Authors
copyright_email: you@example.com
license: Apache-2.0
summary: An InSpec Compliance Profile
version: 0.1.0
depends:
  - name: apache-baseline
    #url: https://github.com/dev-sec/apache-baseline/archive/master.tar.gz
    #git: https://github.com/dev-sec/apache-baseline.git
    #github: dev-sec/apache-baseline
    compliance: admin/apache-baseline
```

`inspec vendor` succeeds.

```
$ inspec vendor issue-2041 --overwrite
Dependencies for profile issue-2041 successfully vendored to issue-2041/vendor
```

But `inspec check` fails.

```
$ inspec check issue-2041
/Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/profile.rb:63:in `for_path': Don't understand inspec profile in /Users/thomaspetchel/.inspec/cache/dc2239efdbe6ed3d638d1f06ceb561891ed03ba0, it doesn't look like a supported profile structure. (RuntimeError)
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/profile.rb:72:in `for_fetcher'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/requirement.rb:106:in `profile'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/dependency_set.rb:65:in `block in each'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/dependency_set.rb:64:in `each'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/dependency_set.rb:64:in `each'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/profile.rb:169:in `load_libraries'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/profile.rb:453:in `load_checks_params'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/profile.rb:447:in `load_params'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/profile.rb:141:in `params'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/profile.rb:307:in `controls_count'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/profile.rb:278:in `check'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/cli.rb:69:in `check'
	from /opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/thor-0.19.1/lib/thor/command.rb:27:in `run'
	from /opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/thor-0.19.1/lib/thor/invocation.rb:126:in `invoke_command'
	from /opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/thor-0.19.1/lib/thor.rb:359:in `dispatch'
	from /opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/thor-0.19.1/lib/thor/base.rb:440:in `start'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/bin/inspec:12:in `<top (required)>'
	from /opt/chefdk/bin/inspec:95:in `load'
	from /opt/chefdk/bin/inspec:95:in `<main>'
```

------

During my investigation, I've seen related failures in `inspec compliance upload` and `inspec archive`. I haven't included repro steps/output here.

### Possible Solutions

🤷‍♂️ 

### Stacktrace

Included in output above.

**Comment by tpetchel** [comment]
Better yet, looks like you don't even need Chef Automate to see the failure. `inspec vendor` should show it.

```
$ inspec vendor my-profile --overwrite
WARN: Unresolved specs during Gem::Specification.reset:
      minitest (~> 5.1)
WARN: Clearing out unresolved specs.
Please report a bug if this causes problems.
/Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/cached_fetcher.rb:15:in `initialize': Could not fetch inspec profile in {:cwd=>"my-profile", :backend=>Inspec::Backend::Class @transport=Train::Transports::Mock::Connection, :name=>"apache-baseline", :github=>"dev-sec/apache-baseline"}. (RuntimeError)
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/requirement.rb:84:in `new'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/requirement.rb:84:in `fetcher'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/requirement.rb:62:in `resolved_source'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/resolver.rb:67:in `block in resolve'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/resolver.rb:59:in `each'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/resolver.rb:59:in `resolve'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/resolver.rb:31:in `resolve'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/dependencies/dependency_set.rb:89:in `vendor'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/profile.rb:393:in `generate_lockfile'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/profile_vendor.rb:58:in `vendor_dependencies'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/profile_vendor.rb:15:in `vendor!'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/base_cli.rb:176:in `vendor_deps'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/lib/inspec/cli.rb:114:in `vendor'
	from /opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/thor-0.19.1/lib/thor/command.rb:27:in `run'
	from /opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/thor-0.19.1/lib/thor/invocation.rb:126:in `invoke_command'
	from /opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/thor-0.19.1/lib/thor.rb:359:in `dispatch'
	from /opt/chefdk/embedded/lib/ruby/gems/2.4.0/gems/thor-0.19.1/lib/thor/base.rb:440:in `start'
	from /Users/thomaspetchel/.chefdk/gem/ruby/2.4.0/gems/inspec-1.31.1/bin/inspec:12:in `<top (required)>'
	from /opt/chefdk/bin/inspec:95:in `load'
	from /opt/chefdk/bin/inspec:95:in `<main>'
```

**Comment by adamleff** [comment]
@tpetchel thanks for the issue!

re `github`: I honestly don't think we ever supported this per my reading of the code. I think this is a documentation error and that section of the docs should be removed. @arlimus @chris-rock keep me honest here, but I can't find any `github` fetcher that would've triggered if someone used it in a `depends` metadata item.

re `git`: this profile fetcher is in need of some work when it comes to vendoring, and you're experiencing one of those rough edges. #1659 is opened and planned for InSpec 2.0 which will correct this issue.

re `compliance`: the error you're getting here is actually a holdover from your `git` testing - you can see that the shasum is the same. You'll need to blow away your inspec cache directory and try again and I believe you'll see it works great (because the compliance fetcher will fetch a .tar.gz and those don't have the issues you experience with the `git` fetcher).

**Comment by jeremymv2** [comment]
@tpetchel @adamleff 

I think there could also be two other things going on here:
https://github.com/chef/inspec/issues/2028
and/or
https://github.com/chef/inspec/pull/2027

Just recently, I've resolved the `inspec check` failure of a vendored profile because it had an empty file entry (related to #2027)

## install latest

chef gem install inspec

## update the version linked in DK

chef gem install appbundle-updater && sudo chef exec appbundle-updater chefdk inspec v1.32.1

## verify works

inspec check d594f6003f6298d576132939f4bc92410130f7e3184163e26235bdd6ed6d7cf3.tar.gz

Location: d594f6003f6298d576132939f4bc92410130f7e3184163e26235bdd6ed6d7cf3.tar.gz
Profile: ge-cis-centos7-v2
Controls: 197
Timestamp: 2017-08-06T08:16:21-04:00
Valid: true

**Comment by chris-rock** [comment]
I agree, we should remove `github` from the documentation since we can just work with the github url

**Comment by kekaichinose** [comment]
This issue has aged, but looks to have been addressed by #2075 and #3286 as there was no follow-up after that.

If this is not the case, please open a new issue, making sure to provide stack trace and repro data that is relevant to the current version of Chef InSpec.

Thank you!
__Chef InSpec Team__
